### DISCUSSION TITLE
Performance for `compile` and `evaluate`

### DISCUSSION THREAD

**Discussion by sylee957** [josdejong/mathjs#3103] [Q&A] [answered]
I'm using mathjs to generate the mathematical functions dynamically for evaluation,
however, I'm getting issues with the performance

In the benchmark:
https://www.measurethat.net/Benchmarks/Show/28847/0/mathjs-performance

```
Native Arrow | Native Arrow x 624 ops/sec ±3.83% (55 runs sampled)
Native Function | Native Function x 545 ops/sec ±4.79% (50 runs sampled)
Native inline | Native inline x 923 ops/sec ±7.12% (9 runs sampled)
Mathjs evaluate | Mathjs evaluate x 10.81 ops/sec ±7.63% (32 runs sampled)
Mathjs compile | Mathjs compile x 34.49 ops/sec ±3.38% (46 runs sampled)
```

Even with `compile`, the performance is 15 times slower than the dynamically generated JS native function with `new Function`.
How do I get the numeric evaluation more faster with mathjs?

**Comment by josdejong** [comment, marked as answer]
Two things here:

1.  You can try do something like:
    ``` 
    var mathfn = math.evaluate('fn(x, y, z) = x^2 + y^2 + z^2')
    ```
    And then run `fn(x, y, z)` in your benchmark
2. The main reason for the slowness is that for evaluation, mathjs uses a heavy weight functions `math.add` and `math.pow` that are capable of handling numbers, BigNumbers, Units, Fractions, Matrices, etc. If you only need `numbers`, you can speed things up a lot by creating a mathjs instance with lightweight functions. To do that, you can use the plain-number implementation via `import {...} from 'mathjs/number'` ([see docs](https://mathjs.org/docs/getting_started.html#es-modules)), or by providing alternative function implementations yourself. Examples:
    - https://mathjs.org/examples/advanced/custom_evaluate_using_import.js.html
    - https://mathjs.org/examples/advanced/custom_evaluate_using_factories.js.html

**Comment by Yuxin-Lou** [comment]
Thank you very much for building such a powerful math library.
I am a js beginner and I hope to find a way to input a string mathematical expression and then output its calculation result. At the same time I am concerned about performance issues. My code won't require calculations on matrices or large numbers, but it will use a lot of Arithmetic functions.
I found a test in the benchmark ( [link to benchmark](https://www.measurethat.net/Benchmarks/Show/17808/0/eval-vs-new-function-vs-mathjs-evaluate-vs-mathjs-chain),) he said eval Performance is higher than evaluate.
But when I executed the following code in the Edge browser console, I found that evaluate is significantly faster than eval. 

What might be the reason for this performance difference? And do you have any suggestions about my work?
```js
var script = document.createElement('script');
script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjs/6.6.1/math.min.js'; 
document.head.appendChild(script);

var average = 0;
for(var test = 0; test<10; test++){
    var startTime = performance.now();
    for (var i = 0; i < 50000; i++) {
        let x = math.evaluate('sqrt(3^2 + 4^2)');
    }
    var endTime = performance.now();
    var elapsedTime = endTime - startTime;
    average += elapsedTime;
}

console.log("mathjs evaluate average result"+ average/10 + " ms");

var average = 0;
for(var test = 0; test<10; test++){
    var startTime = performance.now();
    for (var i = 0; i < 50000; i++) {
        let x = eval('Math.sqrt(3**2 + 4**2)');
    }
    var endTime = performance.now();
    var elapsedTime = endTime - startTime;
    average += elapsedTime;
}

console.log("eval average result"+ average/10 + " ms");
```
result is:
mathjs evaluate average result289.2 ms
eval average result726.7199999988079 ms

**Reply by josdejong** [reply]
First: please test the outcome of your expression `let x = eval('Math.sqrt(3^2 + 4^2)')`, in JavaScript, the ^ operator will do bitwise XOR and and not exponentiation (that is `**` in JavaScript).

Some considerations:

- The mathjs functions support a broad set of data types (numbers, complex numbers, matrices, units, etc). When evaluating `math.add(2, 3)`, it will be slower than a plain `2 + 3` in JavaScript due to type checking overhead. If you don't need all of that, it would be a waste. 
- You can also look into the plain number implementation if you only need numbers, that is faster than the full-fledged functions: https://mathjs.org/docs/custom_bundling.html#numbers-only
- Using JavaScript's `eval` function on dynamic, user provided input comes with security risks, see https://mathjs.org/docs/expressions/security.html

**Reply by Yuxin-Lou** [reply]
After my testing, importing part of mathjs does not seem to significantly speed up calculations.
```sql
import { create, all } from 'mathjs/number'

const math = create(all)

//console.log(math.sqrt(-1)); 
var average = 0;
for(var test = 0; test<50; test++){
    var startTime = performance.now();
    for (var i = 0; i < 50000; i++) {
        let x = math.evaluate('sqrt(3^2 + 4^2)');
    }
    var endTime = performance.now();
    var elapsedTime = endTime - startTime;
    average += elapsedTime;
}

console.log("mathjs evaluate average result"+ average/50 + " ms");
```
![image](https://github.com/josdejong/mathjs/assets/117000057/3f62f092-ba36-428e-ac56-c0ec9690acce)

**Reply by josdejong** [reply]
Using `mathjs/number` can make evaluation faster. You will most likely not see that back in your benchmark because there you're testing parsing _and_ evaluation at the same time, and parsing is the slowest part of the two.

**Reply by Yuxin-Lou** [reply]
Thanks for your patient explanation, that helps a lot.
